### ISSUE TITLE
json module issues with empty dicts and lists

### ISSUE THREAD

**Issue by flenniken** [nim-lang/Nim#6438]
Issue with the json module.

Starting with an empty dict or list then adding to them generates compiler asserts or errors. Here is a simple program showing the issues.

```nim
import json

var myList = %* []
echo myList

var myTable = %* {}
echo myTable # expect {} not []

var myTable2 = %*
  {}
echo myTable2 # expect {} not []

# Error: type mismatch: got (JsonNode) but expected 'JsonNodeObj = object'
# myList[] = %* 5
# echo myList

# Compiler assert
#myTable["hi"] =  %* 42
#echo myTable

# This way works.
var table3 = parseJson("{}")
echo pretty(table3)

table3["hi"] = %* 42
echo table3

table3["meta"] = %* {"filename": "image.dng", "size": 55}
echo table3
```

**Comment by andreaferretti** [comment]
The fact is that the `{a: b, ...}` notation is actually a shortcut for an array of pairs. In particular, `[]` and `{}` are one and the same thing.

**Comment by Yardanico** [comment]
@flenniken there's no such thing as "dict" in Nim.
As @andreaferretti said, {} is just a syntax sugar to array of tuples

**Comment by flenniken** [comment]
I simplified my code. It is true that {} in nim is an array of tuples, but the json representation of "%* {}" should be "{}", shouldn't it?

Test that the json representation of "%* {}" is {}.  It fails.

```nim
import json
var t = %* {}
echo pretty(t)
assert(pretty(t) == "{}")
```

The following generates a compiler assert. Seems like the same bug.

```nim
import json
var t = %* {}
t["hi"] =  %* 42
```

Here is some code from the json module documentation:

```nim
var j2 = %* {"name": "Isaac", "books": ["Robot Dreams"]}
 j2["details"] = %* {"age":35, "pi":3.1415}
 echo j2
```

**Comment by andreaferretti** [comment]
> the json representation of "%* {}" should be "{}", shouldn't it?

The point we are trying to make is that `{}` and `[]` in Nim are one and the same thing. Now, since `{}*` is a macro, it gets to see the AST and could emit a special case for `{}`, is this what you mean?

**Comment by dom96** [comment]
Hrm, we could add a special case here I guess. But is it really worth it?

**Comment by andreaferretti** [comment]
This small thing seems to trip a lot of beginners, so it might be worth

**Comment by pyloor** [comment]
I use `%*` only for batch processing. If i want a empty JsonObject I use `newJsonObject()`. Don't see the needs for a special case. The outcome for it is far to small, newbies have to learn the language. The hint that `[]` and `{}` are the same is far more important.

**Comment by flenniken** [comment]
In Json, [] is not the same as {}.  Json is commonly used to share date
between languages and passing [] will cause problems.


In my case the json dictionary can be empty and it is consumed by Python
and C#. The code starts with an empty "dict" then optionally fills it in. I
noticed the issue when trying to populate the table because it crashes
here.


I am new to nim but I did find the workaround and I will be using it.

>

**Comment by pyloor** [comment]
@flenniken correct. but %* expects a representation in nim and not as json. I understand what your problem is, but the problem is that you mixup json representation and nim representation.

**Comment by andreaferretti** [comment]
@flenniken yes, in JSON the two things are distinct. In Nim, though, the notation `{a: b,c: d,  ...}` as a literal is actually a shortcut for `[(a: b), (c: d)]`. It follows that there is no way to distinguish an empty `{}` from an empty `[]` literal - they are the same value.

The `%*` macro, being a macro, operates on the abstract syntax tree, and thus it could make a special case for `{}` - this is what we are considering here.

As for why the `{}` notation is a shortcut for arrays and not dictionaries. The reason is that arrays are a rather plain data type - a chunk of contiguous memory - while there are various implementations of dictionaries with various tradeoffs. Since Nim is meant to let one write their own performant data structures (as opposed to say Python, where custom implementation of dicts are expected to be much slower than the native ones), it does not choose a particular implementation.

There are dictionary implementations in the stdlib, for instance in the `tables` module, which can be instantiated by something like `{a: b,c: d,  ...}.newTable`

**Comment by pyloor** [comment]
And I am still strongly against a special case for `{}` in JsonModule because it creates a case a common user does not expect. The documentation says it expects a expression. By clicking the source Link in the documentation I can clearly see, that `toJson()` wants a `NimNode`.

`[]` and `{}` are arrays and this has to be even in the json module. It reminds me for the `$` - it converts to a string and this valid for all modules of nim (as far as I know). 

Nim is not Python. In Python a new dict is made by `d = {}` and this converts to a Json Object. And in Nim it is different, period.

ps: sorry for my terrible english :(

**Comment by GULPF** [comment]
If you wan't to be pedantic, `{}` and `[]` is __not__ the same in nim. `{}` is not an empty array, it's an empty set. So the current `*%` macro is not consistent anyway, in fact it specifically rewrites `{}` to `{:}` (which is an empty array). I think it would be much less surprising if it `%* {}` gives an empty JsonObject.

**Comment by Araq** [comment]
I thought this was fixed months ago...

**Comment by Araq** [comment]
Looking at the code, this was always supposed to work this way.

**Comment by GULPF** [comment]
I did some digging, and this was changed due to #4399. It's obvious the original intention of transforming `{}` => `{:}` was to allow empty objects, since using that syntax for an empty array would be bizarre.

I think the best would be to change the `{}` transform to `{}` => `newJObject()`. It's a breaking change, but it's unlikely anybody has ever used `{}` to represent a json array. If that's not an option, the transform should be removed, because it's confusing both from a json perspective and a nim perspective.
